File: train.js

const { Recognizers, Culture } = require('@microsoft/recognizers-text-suite');
const { TextEntityRecognizer, ListEntityRecognizer, RegexEntityRecognizer } = require('botbuilder');
//in this project we want our model to take spoken input, and classify it based on what the user wants to achieve , therefore we're using intent recognition
//we manually provide data samples and their associated intent. it will be used as training data for our model  
// we have to put to mind that improving NLU performance demands that the focus shift from the NLU model to the training data.
// as the project description put us in "customer help center" context=>I narrowed down the intents to three:   shipping policy , return and exchange policy , track norder 
// myModel.addDocument('en', 'hello hey greetings', 'greetings'); => this is bad practice because it takes hello hey greetings as one sentence and no one talks like that so it weakens the model. we have to put real life user utterances also greetingss fit better as entities and not intents
//one of the unnamed tasks u have to do is an analysis for what entities and intents are best based on what you want your model to understand
//here the model does get trained for intents but not for entity extraction! here, entity extraction is not machine learning !it gives us a set of lookup values but it's not data to train the model, in python however that changes with ML libraries like huggingface
myModel.addDocument('en','Do you offer same day shipping?', 'Shipping and handling policy'); 
myModel.addDocument("en","Can you ship to Italy?", "Shipping and handling policy");  
myModel.addDocument("en","How long does delivery take?", "Shipping and handling policy");  //i'm trying to put words variations(package,order,delivery..) to improve the data
myModel.addDocument("en","Can I buy online and pick up my order in store?", "Shipping and handling policy");  
myModel.addDocument("en","What are your shipping options?", "Shipping and handling policy"); 
myModel.addDocument("en","My order arrived damaged, can I get a refund?", " return or exchange policy");  
myModel.addDocument("en","You sent me the wrong item", "return or exchange policy");
myModel.addDocument("en","I want to exchange my item for another colour", "return or exchange policy"); 
myModel.addDocument("en","I ordered something and it wasn't what I expected. Can I return it?", "return or exchange policy");  
myModel.addDocument("en","What's your return policy?", "return or exchange policy");
myModel.addDocument("en","Where's my package?", "Track order");
myModel.addDocument("en","What's my shipping number?", "Track order");
myModel.addDocument("en","Which carrier is my package with?", "Track order"); 
myModel.addDocument("en","Is my package delayed?", "Track order")
 
//now we add our named ,pattern-based entities 
//entities will let my model extract and isolate structured pieces of data from the user's unstructured input
const { TextEntityRecognizer, ListEntityRecognizer, RegexEntityRecognizer } = require('botbuilder');

// Create recognizers
const textRecognizer = new TextEntityRecognizer();
const listRecognizer = new ListEntityRecognizer();
const regexRecognizer = new RegexEntityRecognizer();
//here our entity varibale 'subscription_plan' encapsulates three variations of subscription! what I'm trying to do is make my model performant and as i said the focus needs to shift from the model itself to the data !
listRecognizer.addInstance('subscription_plan', 'basic', ['basic', 'basic plan', 'monthly plan']);
listRecognizer.addInstance('subscription_plan', 'premium', ['premium', 'premium plan', 'gold plan', 'yearly plan']);
listRecognizer.addInstance('subscription_plan', 'trial', ['trial', 'free trial', 'trial period']);
listRecognizer.addInstance('delivery_status',null,'eng',['shipped','processing','out for delivery','delivered','delayed']);
listRecognizer.addInstance('service_type',null,'eng',['refund',' exchange', 'technical support', 'account help']);
listRecognizer.addInstance('return_reason', 'damaged', ['damaged', 'broken', 'defective', 'arrived damaged']);
listRecognizer.addInstance('return_reason', 'wrong_item', ['wrong item', 'incorrect item', 'sent the wrong item']);
listRecognizer.addInstance('return_reason', 'not_as_expected', ['not what I expected', "doesn't match description", 'different than pictured']);
listRecognizer.addInstance('return_reason', 'size_issue', ['wrong size', 'too small', 'too large', 'size issue']);

listRecognizer.addInstance('request_type', 'refund', ['refund', 'I want a refund', 'refund request']);
listRecognizer.addInstance('request_type', 'exchange', ['exchange', 'I want to exchange', 'replace item']);
listRecognizer.addInstance('request_type', 'cancel_order', ['cancel order', 'i want to cancel', 'cancel my purchase']);
listRecognizer.addInstance('request_type', 'status_update', ['order status', 'where is my order', 'tracking update']);

listRecognizer.addInstance('greeting_word', 'hello', ['hello', 'hi', 'hey', 'good morning', 'good evening']);

// Regex entities (pattern-based)
regexRecognizer.addRegexEntity('order_number','eng', /\b#[0-9]{4}\b/i);
regexRecognizer.addRegexEntity('location', /\b[A-Z][a-z]+(?:\s[A-Z][a-z]+)*\b/);
regexRecognizer.addRegexEntity('date', /\b(?:\d{4}[-/]\d{1,2}[-/]\d{1,2}|\d{1,2}[\/-]\d{1,2}[\/-]\d{2,4}|today|tomorrow|yesterday|next week|next month)\b/gi);
regexRecognizer.addRegexEntity('price_amount', /\b(?:\$|€|£)\s?\d+(?:\.\d{2})?|\b\d+(?:\.\d{2})?\s?(USD|EUR|GBP)\b/gi);
regexRecognizer.addRegexEntity('discount_code', /\b[A-Z0-9]{5,15}\b/);

//Now these recognizers can be used in the bot's answers!
module.exports = { textRecognizer, listRecognizer, regexRecognizer };

async function main(){
    try{
    await myModel.train()
    await myModel.save()
    const response =  await myModel.process('en', 'I have a complaint about my order you gave me the wrong size');
    console.log("reponse est: ",response); 

    }catch(e){
        console.error(e);
    } 
}    
main();

File: service-python/train.py
#in this project we want our model to take spoken input, and classify it based on what the user wants to achieve , therefore we're using intent recognition
#we manually provide data samples and their associated intent. it will be used as training data for our model  
# we have to put to mind that improving NLU performance demands that the focus shift from the NLU model to the training data.
#as the project description put us in "customer help center" context=>I narrowed down the intents to three:   shipping policy , return and exchange policy , track norder 
#['hello hey greetings'] => this is bad practice because it takes hello hey greetings as one sentence and no one talks like that so it weakens the model. we have to put real life user utterances also greetingss fit better as entities and not intents
#one of the unnamed tasks u have to do is an analysis for what entities and intents are best based on what you want your model to understand
#here the model does get trained for intents but not for entity extraction! here, entity extraction is not machine learning !it gives us a set of lookup values but it's not data to train the model,  however that changes with ML libraries like huggingface
from processor import preprocessing
from vectorize import vectorizing
import json 
import pickle
import sklearn 
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.calibration import CalibratedClassifierCV

sample=[]
target=[]
with open('dataset.json','r',encoding='utf8') as file:
    train_data=json.load(file)
for intent in train_data['intents']:
    for value in intent['samples'] :
        clean_sample=preprocessing(value) #we preprocess the value and take it as raw text. the preprocessing() function will return a string for each value.
        sample.append(clean_sample) #now we encapsulate the strings into one list 'sample'.
        #tfidf later on input is a list directly made of strings not a list made of lists. One of the unnamed tasks i noticed is that i have to put in mind what each function takes as input and what's the type it returns
        target.append(intent['intent']) #the target needs to be inside the second loop(same number of entries to X and y) bc each X needs to be associated to its own label otherwise we'd have Null values in y

X_train,X_test,y_train,y_test=train_test_split(sample,target,train_size=0.8,random_state=42)
dict=vectorizing(X_train,X_test) #les données de training et test transformée en matrice
X_train_matrix=dict['key1']
X_test_matrix=dict['key2']

#modele régression logistique
#log_reg=LogisticRegression(max_iter=900)
#log_reg.fit(X_train_matrix,y_train) #y n'a pas besoin d'etre transformée , elle peut etre en string et après le modèle la transforme 
#y_pred=log_reg.predict(X_test_matrix)

#precision=precision_score(y_test,y_pred,average=None)
#print(f"Précision pour régression logistique: {precision}")
#accuracy=accuracy_score(y_test,y_pred)
#print(f"accuracy: {accuracy}")
#print(confusion_matrix(y_test,y_pred))

#modele Linear Support Vector Classifier
classifier = SVC(kernel="linear")
calibrated_classifier=CalibratedClassifierCV(classifier)
calibrated_classifier.fit(X_train_matrix, y_train)
y_pred=calibrated_classifier.predict(X_test_matrix)
precision=precision_score(y_test,y_pred,average=None)
print(f"Précision pour svm: {precision}")
accuracy=accuracy_score(y_test,y_pred)
print(f"accuracy: {accuracy}")
print(confusion_matrix(y_test,y_pred))
if(hasattr(calibrated_classifier, "predict_proba") == True):
    print('true')
else:
    print('false')

#modele multinomial naive bayes
#modele=MultinomialNB()
#modele.fit(X_train_matrix,y_train)
#y_pred=modele.predict(X_test_matrix)
#precision=precision_score(y_test,y_pred,average=None)
#print(f"Précision pour multinomial NB: {precision}")
#accuracy=accuracy_score(y_test,y_pred)
#print(f"accuracy: {accuracy}")
#print(confusion_matrix(y_test,y_pred))

#apres comparaison entre les 'confusion matrix' et les 'metrics' des trois modèles, je choisis le SVM . 

with open('model.pkl','wb') as f :
   pickle.dump(calibrated_classifier,f) #we pickle the model to not build a new one each time , we just unpickle it out of the disk! 